lib/exla/client.ex

```diff
@@ -5,7 +5,7 @@ defmodule Exla.Client do
   @enforce_keys [:ref]
   defstruct [:ref]
 
-  def create_client(options = %LocalClientOptions{}) do
+  def create_client(options \\ %LocalClientOptions{}) do
     case Exla.NIF.get_or_create_local_client(
```

Comment:
Btw, do we have a `create_local_client` API that doesn't return an existing one? Also, if we have different client configurations, is XLA guaranteeing each of them return a different client instance?

Reply:
We do not. The current one wraps the GetOrCreate XLA function. I was looking at this earlier and realized that if we continue calling this method it will return new references to the same object on the XLA side. So unless we explicitly create a new one it just keeps returning different references to the same thing

Reply:
>  So unless we explicitly create a new one it just keeps returning different references to the same thing

Does XLA allow us to do that? Also what happens if I call this function with different options, are they different clients on the C++ side?

In theory it is ok if we are using the same... as long as it is thread safe?

Reply:
We can explicitly create a new one if we specify a different platform. For example if I have a CUDA client and a CPU client, I can create two clients for interacting with each locally. But, if I tried to create 2 clients on my host device with a different `num_replicas` option, it would ignore the options and return the already instantiated client on the host device.

There's also different types of clients that need to be handled at some points like: gRPC clients and compile-only clients. It seems, though, that local clients are the most common so I'm going to only focus on those until we have more built out.

It's not explicitly marked as being thread-safe and going through the code it doesn't seem to be. The local client is supposed to be a singleton object wrapping a platform/service, and trying to use two on one platform can create some issues especially when dealing with some of the mutable attributes you can access through the client.

Reply:
I see, so after-all we may want to model this using Elixir's Registry + GenServer, but I think we can wait with moving ahead with this. I think for now I would just leave a TODO:

```suggestion
    # TODO: Rename this function to get_local_client. It is a singleton,
    # non-thread-safe resource in XLA so we need to mimic the same
    # in Elixir. We should also have distinct steps for configuring and for
    # getting it.
    case Exla.NIF.get_or_create_local_client(
```